File: ContiguousArray.py
class Solution(object):
    def findMaxLength(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        a = {}
        
        
        a[0] = -1
        
        total = 0
        
        maxLength = 0
        
        for i in range(len(nums)):
            
            if(nums[i] == 0):
                
                total = total + 1
            
            else:
                
                total = total - 1
                
            
            if total in a:
                
                length = i - a[total]
                maxLength = max(maxLength, length)
                
            else:
                
                a[total] = i
        
        return maxLength

File: ContiguousArray.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) 
    {
        int count = 0;
        
        int maxLength = 0;
        
        unordered_map<int, int> map;
        
        map[0] = 0;
        
        int size = nums.size();
        
        for(int i = 0; i < size; i++)
        {
            if(nums[i] == 0)
            {
                count = count + 1;
            }
            
            if(nums[i] == 1)
            {
                count = count - 1;
            }
            
            if(map.find(count) != map.end())
            {
                int length = i + 1 - map[count];
                maxLength = max(maxLength, length);
            }
            
            else
            {
                map[count] = i + 1;
            }
        }
        
        return maxLength;
        
    }
};